webui/src/App.js

```diff
@@ -21,86 +21,168 @@ import {
+// Maximum number of historical sensor values to retain
+const max_size = 1000;
 
-// Keep track of the current thresholds fetched from the backend.
-// Make it global since it's used by many components.
-let kCurThresholds = new Array(num_panels).fill(0);
+// Reference to history of current sensor values.
+// The values are stored in a mutable array in a ref so that they are not
+// subject to the React render cycle, for performance reasons.
+// The Context is to make them easy to access from nested components.
+const CurValuesRefContext = React.createContext();
 
-// A history of the past 'max_size' values fetched from the backend.
-// Used for plotting and displaying live values.
-// We use a cyclical array to save memory.
-let kCurValues = [new Array(num_panels).fill(0)];
-const max_size = 1000;
-let oldest = 0;
+// Returned `defaults` property will be undefined if the defaults are loading or reloading.
+// Call `reloadDefaults` to clear the defaults and reload from the server.
+function useDefaults() {
+  const [defaults, setDefaults] = useState();
 
-var ws;
-const wsCallbacks = {};
-const wsQueue = [];
+  const reloadDefaults = useCallback(() => setDefaults(undefined), [setDefaults]);
 
-function connect() {
-  ws = new WebSocket('ws://' + window.location.host + '/ws');
+  // Load defaults at mount and reload any time they are cleared.
+  useEffect(() => {
+    let cleaningUp = false;
+    let timeoutId = 0;
 
-  ws.addEventListener('open', function(ev) {
-    while (wsQueue.length > 0 && ws.readyState === 1) {
-      let msg = wsQueue.shift();
-      ws.send(JSON.stringify(msg));
+    const getDefaults = () => {
+      clearTimeout(timeoutId);
+      fetch('/defaults').then(res => res.json()).then(data => {
+        if (!cleaningUp) {
+          setDefaults(data);
+        }
+      }).catch(reason => {
+        if (!cleaningUp) {
+          timeoutId = setTimeout(getDefaults, 1000);
+        }
+      });
     }
-  });
 
-  ws.addEventListener('error', function(ev) {
-    ws.close();
-  });
+    if (!defaults) {
+      getDefaults();
+    }
+
+    return () => {
+      cleaningUp = true;
+      clearTimeout(timeoutId);
+    };
+  }, [defaults]);
+
+  return { defaults, reloadDefaults };
+}
 
-  ws.addEventListener('close', function(ev) {
-    setTimeout(connect, 1000);
+// returns { curValuesRef, emit, isWsReady, wsCallbacksRef }
+// curValuesRef tracks values as well as thresholds.
+// Use emit to send events to the server.
+// isWsReady indicates that the websocket is connected and has received
+// its first values reading from the server.
+// onCloseWs is called when the websocket closes for any reason,
+// unless the hooks is already doing effect cleanup.
+// The expectation is that onCloseWs is used to reload default thresholds
+// and provide new ones to trigger a new connection.
+function useWsConnection({ defaults, onCloseWs }) {
+  const [isWsReady, setIsWsReady] = useState(false);
+
+  const curValuesRef = useRef({
```

Comment:
"curValuesRef" includes kCurValues and kCurThresholds, but it would be relatively straightforward to put them in different refs if desired. If they stay here, I'm not sure about the name.

Incidentally, what is the `k` prefix for?

Reply:
The disambiguation I've been trying to use is that `Values` refers to the sensor readings and `Thresholds` are the trigger levels.

Combining is fine but I would want to rename it. Since it encapsulates all the data needed for the WebUI, maybe something like `webUIDataRef`

Reply:
That naming sounds good. I like the `Ref` suffix because if you want to give a name to the current value it can easily be `const webUIData = webUIDataRef.current`.

Reply:
SGTM!